mcpserver/tools/util.go

```diff
@@ -44,7 +76,49 @@ func fetchFileData(filespec string) ([]byte, error) {
 
 	// Handle as file path
 	cleanPath := filepath.Clean(filespec)
-	file, err := os.Open(cleanPath)
+
+	// Use data directory as base for file operations
+	dataDir, err := GetDataDirectoryInternal()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get data directory: %w", err)
+	}
+
+	// Ensure data directory exists
+	if dirErr := EnsureDataDirectory(); dirErr != nil {
+		return nil, fmt.Errorf("failed to create data directory: %w", dirErr)
+	}
+
+	// Reject absolute paths upfront
+	if filepath.IsAbs(cleanPath) {
+		return nil, fmt.Errorf("absolute paths not allowed: %s", filespec)
+	}
+
+	// Build and resolve paths for containment verification
+	requestedPath := filepath.Join(dataDir, cleanPath)
+	absDataDir, err := filepath.Abs(dataDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve data directory: %w", err)
+	}
+
+	absRequestedPath, err := filepath.Abs(requestedPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve requested path: %w", err)
+	}
+
+	// Use filepath.Rel to verify the requested path is within data directory
+	rel, err := filepath.Rel(absDataDir, absRequestedPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine path relationship: %w", err)
+	}
+
+	// If rel starts with ".." or is "..", the path is outside data directory
+	if strings.HasPrefix(rel, ".."+string(filepath.Separator)) || rel == ".." {
+		return nil, fmt.Errorf("path outside data directory: %s", filespec)
+	}
+
+	filePath := absRequestedPath
+
+	file, err := os.Open(filePath)
```

Comment:
We can probably simplify and make this code more secure with the new https://pkg.go.dev/os#Root in the os library. There is a blog post about it here: https://go.dev/blog/osroot